### ISSUE TITLE
HDF5Dataset

### ISSUE THREAD

**Issue by hs199510** [wolny/pytorch-3dunet#31]
hi,I want to know how your h5dataset is made.Is hdf5.py only used to parse  files, like random_label3D.h5?

**Comment by lpengc** [comment]
@wolny , I also want to know how should I create the .h5 file, by the way, my data format is .nii .

**Comment by zz10001** [comment]
Hi，@lpengc 
have you sovle it?can you give some advice

**Comment by StuckinPhD** [comment]
You need to create your own ``hdf5`` files for your own dataset. This is very simple to achieve in python, use a library like ``h5py`` and make sure to label your dataset as ``raw`` for the sample volumes and ``label`` for its corresponding ground truth volume.

For example this below little script should generate your hdf5 file needed for this network.

```
import h5py


save_path = os.path.join('train_hdf5', 'train_file_.h5')

hf = h5py.File(save_path, 'w')
hf.create_dataset('raw', data=volume, chunks=True, compression='gzip')
hf.create_dataset('label', data=label, chunks=True, compression='gzip')
hf.close()
```

where ``volume`` is your sample and ``label`` is the ground truth for that sample. You have to read these in as numpy arrays or whatever before writing them to h5 format. 

 Then give the location of this file in the ``train_config_.yaml`` file.

```
train_path:
    - '/train_hdf5/train_file_1.h5'
    - '/train_hdf5/train_file_2.h5'
```

**Comment by zz10001** [comment]
Thanks for your quick reply,i will have a try latter.may you have a nice day!
Best,
ming

**Comment by zzz123xyz** [comment]
Hi @StuckinPhD @wolny could you provide some more information about how to create the hdf5 file from nifti file (.nii). I have the same issues here. 
    Given I have a folder including samples (.nii) files, I have another folder including grand truth segmentation labels (.nii). All of the samples and labels are 3D. the files structure is as follow:
```
  folder_samples
    |
    |--sample1.nii
    |--sample2.nii
    |--sample3.nii

  folder_labels
    |
    |--label1.nii
    |--label2.nii
    |--label3.nii
```
So how can I construct the structure in the hdf5 file (.h5)?

As you suggest 

```
import h5py
import os

save_path = os.path.join('train_hdf5', 'train_file_.h5')

hf = h5py.File(save_path, 'w')
hf.create_dataset('raw', data=volume, chunks=True, compression='gzip')
hf.create_dataset('label', data=label, chunks=True, compression='gzip')
hf.close()
```
and the sample file provided in the resources/ folder 
https://github.com/wolny/pytorch-3dunet/blob/master/resources/random_label3D.h5
I had checked in a HDF viewer.

I suppose the structure in the HDF file (.h5) looks like:
```
  random_label3D.h5
    |
    |--label
    |--raw

```

I am just very confused about the structure of the created hdf5 file.
Should all the samples be concatenated in the first dimension and thrown into the |--raw, and so do the labels be thrown into the  |--label 
or should it be like the following structure:

```
dataset_train.h5
    |
    |--sample1/
        |--label
        |--raw
    |--sample2/
        |--label
        |--raw
    ....
```
Please provide more information or code samples for this part. Thanks very much

**Comment by ww-Jingle** [comment]
> Hi @StuckinPhD @wolny could you provide some more information about how to create the hdf5 file from nifti file (.nii). I have the same issues here.
> Given I have a folder including samples (.nii) files, I have another folder including grand truth segmentation labels (.nii). All of the samples and labels are 3D. the files structure is as follow:
> 
> ```
>   folder_samples
>     |
>     |--sample1.nii
>     |--sample2.nii
>     |--sample3.nii
> 
>   folder_labels
>     |
>     |--label1.nii
>     |--label2.nii
>     |--label3.nii
> ```
> 
> So how can I construct the structure in the hdf5 file (.h5)?
> 
> As you suggest
> 
> ```
> import h5py
> import os
> 
> save_path = os.path.join('train_hdf5', 'train_file_.h5')
> 
> hf = h5py.File(save_path, 'w')
> hf.create_dataset('raw', data=volume, chunks=True, compression='gzip')
> hf.create_dataset('label', data=label, chunks=True, compression='gzip')
> hf.close()
> ```
> 
> and the sample file provided in the resources/ folder
> https://github.com/wolny/pytorch-3dunet/blob/master/resources/random_label3D.h5
> I had checked in a HDF viewer.
> 
> I suppose the structure in the HDF file (.h5) looks like:
> 
> ```
>   random_label3D.h5
>     |
>     |--label
>     |--raw
> ```
> 
> I am just very confused about the structure of the created hdf5 file.
> Should all the samples be concatenated in the first dimension and thrown into the |--raw, and so do the labels be thrown into the |--label
> or should it be like the following structure:
> 
> ```
> dataset_train.h5
>     |
>     |--sample1/
>         |--label
>         |--raw
>     |--sample2/
>         |--label
>         |--raw
>     ....
> ```
> 
> Please provide more information or code samples for this part. Thanks very much

According to the function get_train_loaders in hdf5.py file, you may create a group .h5 files for storing raw and label data of a group .nii files. 
```
train_path:
    - '/train_hdf5/train_file_1.h5'
    - '/train_hdf5/train_file_2.h5'
    -...
```
A sample.nii and a label.nii file should be constructed together to get a train_file.h5 file.

**Comment by wolny** [comment]
Hi @zzz123xyz,

HDF5 is the only format supported currently and you have to convert the raw/label files that you have to the HDF5 format on your own as explained by @StuckinPhD (https://github.com/wolny/pytorch-3dunet/issues/31#issuecomment-510124132).
